### PR TITLE
KWS-2340: AppRouter + Views conversion to ViewController

### PR DESCRIPTION
**Pull request by Tbruni85**
## JIRA
https://jira.superawesome.tv/browse/KWS-2340

## DESCRIPTION
Implementing:
* *AppRouter*
* *RoutingPath*
in the UI lib.

**Refactor**
Converting previously classes of type ViewController to View and corrisponding ViewController in the Example app.

## SCREENSHOTS
n/a

### CODE REVIEW

File: Example/PopJamUI/CreatePasswordView.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+//  Created by Tiziano Bruni on 09/02/2018.
+//
+
+import Foundation
+import SAProtobufs
+import PopJamSDK
+import PopJamUI
+import PureLayout
+
+public class CreatePasswordViewController: UIViewController, CreatePasswordViewDelegate  {
```
Comment (me2k) [inline Example/PopJamUI/CreatePasswordView.swift]: I would move the `CreatePasswordViewDelegate` into an extension at the bottom of the file to separate the functions from the rest of the UIViewController code.

File: Example/PopJamUI/InitialViewController.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+//
+//  Created by Tiziano Bruni on 29/01/2018.
+//
+
+import Foundation
+import SAProtobufs
+import PopJamUI
+import PureLayout
+
+public class InitialViewController: UIViewController, InitialViewDelegate {
```
Comment (me2k) [inline Example/PopJamUI/InitialViewController.swift]: Move `InitialViewDelegate` into an extension

File: Example/PopJamUI/InitialViewController.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required public init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+    }
+    
+    override public func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        // Hide the navigation bar on the this view controller
```
Comment (me2k) [inline Example/PopJamUI/InitialViewController.swift]: This comment isn't needed as the line of code below it describes your intent just as well

File: Example/PopJamUI/ViewController.swift [hunk 4]
```diff
@@ -130,18 +130,17 @@ class ViewController: UIViewController,
...
     
     func initInitialView() {
         
         let initialViewModel = InitialViewModel()
-        let initialView = InitialView.init(withViewModel: initialViewModel, config: self.config)
-        initialView.delegate = self
+        let initialViewController = InitialViewController.init(withViewModel: initialViewModel, config: config)
         
-        self.present(initialView, animated: true, completion: nil)
+        self.present(initialViewController, animated: true, completion: nil)
```
Comment (me2k) [inline Example/PopJamUI/ViewController.swift]: delete self as it is not needed here ;)

File: Example/PopJamUI/ViewController.swift [hunk 5]
```diff
@@ -160,9 +159,9 @@ class ViewController: UIViewController,
         let dataProvider = CreatePasswordDataProvider.init(withUsername: username,
                                                             authService: authService!,
                                                             config: self.config)
-        let createPasswordView = CreatePasswordView.init(withDataProvider: dataProvider)
+        let createPasswordViewController = CreatePasswordViewController.init(withDataProvider: dataProvider)
         
-        self.navigationController?.pushViewController(createPasswordView, animated: true)
+        self.navigationController?.pushViewController(createPasswordViewController, animated: true)
```
Comment (me2k) [inline Example/PopJamUI/ViewController.swift]: remove self here too

File: PopJamUI/Classes/RoutingPaths/AppRouter.swift [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    
+    // MARK: PushView Notification handler
+    @objc func decodeRouteRequest(notification: Notification) -> Void {
+        
+        guard let userInfo = notification.userInfo else {
+            print("No userInfo found in notification")
+            return
+        }
+        
+        let path = userInfo["path"] as? RoutingPath
```
[the remark this answers is not in the document]
Comment (Tbruni85) [reply PopJamUI/Classes/RoutingPaths/AppRouter.swift]: class not still finished

File: PopJamUI/Classes/RoutingPaths/AppRouter.swift [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        
+    }
+    
+    private func pathResolver(path: RoutingPath) -> UIViewController? {
+        
+        
+        let anyobjectype : AnyObject.Type = NSClassFromString("Router_Test_App." + path.destinationView.rawValue)!
+        
+
```
[the remark this answers is not in the document]
Comment (Tbruni85) [reply PopJamUI/Classes/RoutingPaths/AppRouter.swift]: class not still finished

File: PopJamUI/Classes/InitialView/Views/InitialView.swift [hunk 8]
```diff
@@ -135,17 +124,17 @@ public class InitialView: UIViewController, UITextViewDelegate {
...
         emoji_image.autoPinEdge(toSuperviewEdge: .trailing, withInset: -12.0)
         
-        self.getStartedButton.autoAlignAxis(toSuperviewAxis: .vertical)
-        self.getStartedButton.autoPinEdge(.top, to: .bottom, of: emoji_image, withOffset: 26.0)
-        self.getStartedButton.autoPinEdge(toSuperviewEdge: .leading, withInset: 28.0)
-        self.getStartedButton.autoPinEdge(toSuperviewEdge: .trailing, withInset: 28.0)
-        self.getStartedButton.autoSetDimension(.height, toSize: 40.0)
+        getStartedButton.autoAlignAxis(toSuperviewAxis: .vertical)
+        getStartedButton.autoPinEdge(.top, to: .bottom, of: emoji_image, withOffset: 26.0)
+        getStartedButton.autoPinEdge(toSuperviewEdge: .leading, withInset: 28.0)
```
[the remark this answers is not in the document]
Comment (Tbruni85) [reply PopJamUI/Classes/InitialView/Views/InitialView.swift]: I will update constraint setting style when we have more time

File: PopJamUI/Classes/RoutingPaths/AppRouter.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            
+            let navController = UINavigationController.init(rootViewController: view!)
+            originView.present(navController, animated: true, completion: nil)
+        }
+        
+    }
+    
+    private func pathResolver(path: RoutingPath) -> UIViewController? {
+        
+
```
[the remark this answers is not in the document]
Comment (Tbruni85) [reply PopJamUI/Classes/RoutingPaths/AppRouter.swift]: class not still finished

File: PopJamUI/Classes/RoutingPaths/AppRouter.swift [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        
+        
+        let anyobjectype : AnyObject.Type = NSClassFromString("Router_Test_App." + path.destinationView.rawValue)!
+        
+        
+        switch path.destinationView {
+        case .loginView:
+//            let nsobjectype : LoginView.Type = anyobjectype as! LoginView.Type
+//            let rec: LoginView = nsobjectype.init
+//            return rec
```
[the remark this answers is not in the document]
Comment (Tbruni85) [reply PopJamUI/Classes/RoutingPaths/AppRouter.swift]: class not still finished

File: PopJamUI/Classes/RoutingPaths/AppRouter.swift [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+            
+            print("loginView")
+        case .createUserNameView:
+            
+//            let nsobjectype : CreateUserNameView.Type = anyobjectype as! CreateUserNameView.Type
```
[the remark this answers is not in the document]
Comment (Tbruni85) [reply PopJamUI/Classes/RoutingPaths/AppRouter.swift]: class not still finished

File: PopJamUI/Classes/RoutingPaths/RoutingPath.swift [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+public struct RoutingPath {
+    
+    var originView: UIViewController
+    var destinationView: ViewType
+    var currentNavController: UINavigationController?
+    var pathType: PathType
+    var needNavigationController: Bool
+    var parameters: [String:Any]?
+    
+
```
[the remark this answers is not in the document]
Comment (Tbruni85) [reply PopJamUI/Classes/RoutingPaths/RoutingPath.swift]: class not still finished

File: Example/PopJamUI/ViewControllers/CreatePasswordViewController.swift [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+public class CreatePasswordViewController: UIViewController, CreatePasswordViewDelegate  {
+    
+    // UI components
+    var createPasswordView: CreatePasswordView!
+    
+    // Logic components
+    private var dataProvider: CreatePasswordDataProvider!
+    
+
```
Comment (me2k) [inline Example/PopJamUI/ViewControllers/CreatePasswordViewController.swift]: Delete this new line

File: PopJamUI/Classes/RoutingPaths/AppRouter.swift [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+            if let navC = path?.currentNavController! {
+                navController = navC
+            } else {
+                return
+            }
+        }
+        
+        switch pathType {
+        case .push:
+            if (path?.needNavigationController)! {
```
Comment (devgabrielcoman) [inline PopJamUI/Classes/RoutingPaths/AppRouter.swift]: I know it's still "in progress" but why are these things explicitly unwrapped? that shouldn't be the case, it'll generate bundles of problems down the line.

**Pull request merged**